<a href="https://colab.research.google.com/github/KomisD/Neural-Networks-Classification/blob/main/NN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Multi-Class Classification using Multilayer Perceptron Neural Network




##  Introduction


 dataset: https://www.kaggle.com/datasets/nelgiriyewithana/billionaires-statistics-dataset

###- Project Description
This project focuses on the analysis and classification of a unique dataset containing detailed statistics about the world's billionaires.

The classifications to be performed are threefold:


*   **Wealth Category Classification**: Categorize billionaires into different wealth brackets based on their net worth.
*   **Industry Classification**: Classify billionaires according to the primary industry of their businesses.
*   **Self-Made vs. Inherited Wealt**h: Determine whether the wealth of each billionaire is self-made or inherited.


##  Dataset Selection and Preprocessing

###Importing the libraries

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt

###Importing the Dataset

In [ ]:
dataset = pd.read_csv('Billionaires Statistics Dataset.csv')

dataset.iloc[:, 26] = dataset.iloc[:, 26].str.replace('$', '')
dataset.iloc[:, 26] = dataset.iloc[:, 26].str.replace(',', '')

# Remove rows where any of the data is missing
dataset = dataset.dropna()

# category is the same with industries (exact same)
dataset.drop(columns=dataset.columns[2], inplace=True)

#Rank is not helping in any of the models (also in final worth guess would be a cheat)
dataset.drop(columns=dataset.columns[0], inplace=True)

#personName is not helping in any of the presictions
dataset.drop(columns=dataset.columns[1], inplace=True)

#city is not something I will try to predict or use
dataset.drop(columns=dataset.columns[3], inplace=True)

#city is not something I will try to predict or use
dataset.drop(columns=dataset.columns[3], inplace=True)

#too many different sources to help indicate patterns (difficault with one hot encoding)
dataset.drop(columns=dataset.columns[5], inplace=True)

#droping the birth date since is captured by the birth year, birth month and birth day
dataset.drop(columns=dataset.columns[8], inplace=True)

#Names are not helping in any categorize
dataset.drop(columns=dataset.columns[8], inplace=True)
dataset.drop(columns=dataset.columns[8], inplace=True)

#Title have too many missing values
dataset.drop(columns=dataset.columns[8], inplace=True)

#date of the update of the data unimporntant
dataset.drop(columns=dataset.columns[8], inplace=True)

#the same as the city, I will focus only in the country and not state or residence state
dataset.drop(columns=dataset.columns[8], inplace=True)
dataset.drop(columns=dataset.columns[8], inplace=True)


<ipython-input-294-ee3f218a2899>:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dataset.iloc[:, 26] = dataset.iloc[:, 26].str.replace('$', '')
<ipython-input-294-ee3f218a2899>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset.drop(columns=dataset.columns[2], inplace=True)
<ipython-input-294-ee3f218a2899>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset.drop(columns=dataset.columns[0], inplace=True)
<ipython-input-294-ee3f218a2899>:16: SettingWit

In [ ]:
print(dataset.iloc[0, :].values)

[180000 51.0 'United States' 'Automotive' 'United States' True 'D' 'M'
 1971.0 6.0 28.0 117.24 7.5 '21427700000000 ' 88.2 101.8 78.5 9.6 36.6
 328239523.0 37.09024 -95.712891]


In [ ]:
X_Wealth_Classification = dataset.iloc[:, 1:].values
y_Wealth_Classification = dataset.iloc[:, 0].values.copy()

In [ ]:
print(X_Wealth_Classification[0, :])

[51.0 'United States' 'Automotive' 'United States' True 'D' 'M' 1971.0 6.0
 28.0 117.24 7.5 '21427700000000 ' 88.2 101.8 78.5 9.6 36.6 328239523.0
 37.09024 -95.712891]


In [ ]:
#categorize the bilioners into ultra wealth, median and low
for i in range(len(y_Wealth_Classification)):
    if y_Wealth_Classification[i] > 60000:
        y_Wealth_Classification[i] = 0
    elif y_Wealth_Classification[i] > 10000:
        y_Wealth_Classification[i] = 1
    else:
        y_Wealth_Classification[i] = 2

In [ ]:
X_Industry_Classification = dataset.iloc[:, :].values
X_Industry_Classification = np.delete(X_Industry_Classification, 3, axis=1)
y_Industry_Classification = dataset.iloc[:, 3].values

In [ ]:
print(X_Industry_Classification[0, :])

[180000 51.0 'United States' 'United States' True 'D' 'M' 1971.0 6.0 28.0
 117.24 7.5 '21427700000000 ' 88.2 101.8 78.5 9.6 36.6 328239523.0
 37.09024 -95.712891]


In [ ]:
# Find the unique categories and the count of each
unique_categories, counts = np.unique(y_Industry_Classification, return_counts=True)

# Number of unique categories
num_unique_categories = len(unique_categories)

print(num_unique_categories)

16


***Here I didn't applied OneHotEncoding cause both neural and KNN are sensitive algoriths to the dimentionality and i have 16 different industries ***

In [ ]:
X_SelfVsInherited_Classification = dataset.iloc[:, :].values
X_SelfVsInherited_Classification = np.delete(X_SelfVsInherited_Classification, 5, axis=1)
y_SelfVsInherited_Classification = dataset.iloc[:, 5].values

In [ ]:
print(X_SelfVsInherited_Classification[0,:])

[180000 51.0 'United States' 'Automotive' 'United States' 'D' 'M' 1971.0
 6.0 28.0 117.24 7.5 '21427700000000 ' 88.2 101.8 78.5 9.6 36.6
 328239523.0 37.09024 -95.712891]


###Encoding the Indipendent Variable

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

#didn't use oneHotEncoder (tryied but there are way too many countries and industries)
label_encoder = LabelEncoder()
X_Wealth_Classification[:, 2] = label_encoder.fit_transform(X_Wealth_Classification[:, 2])
X_Wealth_Classification[:, 3] = label_encoder.fit_transform(X_Wealth_Classification[:, 3])
X_Wealth_Classification[:, 4] = label_encoder.fit_transform(X_Wealth_Classification[:, 4])

X_Wealth_Classification[:, 5] = label_encoder.fit_transform(X_Wealth_Classification[:, 5])
X_Wealth_Classification[:, 6] = label_encoder.fit_transform(X_Wealth_Classification[:, 6])
X_Wealth_Classification[:, 1] = label_encoder.fit_transform(X_Wealth_Classification[:, 1])

#Inductry classification
y_Industry_Classification = label_encoder.fit_transform(y_Industry_Classification)
X_Industry_Classification[:, 2] = label_encoder.fit_transform(X_Industry_Classification[:, 2])
for i in range(2,7):
  X_Industry_Classification[:, i] = label_encoder.fit_transform(X_Industry_Classification[:, i])

#Self-Made vs. Inherited Wealth
y_SelfVsInherited_Classification = label_encoder.fit_transform(y_SelfVsInherited_Classification)
for i in range(2,7):
  X_SelfVsInherited_Classification[:, i] = label_encoder.fit_transform(X_SelfVsInherited_Classification[:, i])

In [ ]:
print(X_Wealth_Classification[0,:])

[51.0 0 0 7 1 0 1 1971.0 6.0 28.0 117.24 7.5 '21427700000000 ' 88.2 101.8
 78.5 9.6 36.6 328239523.0 37.09024 -95.712891]


In [ ]:
print(X_Industry_Classification[0,:])

[180000 51.0 0 7 1 0 1 1971.0 6.0 28.0 117.24 7.5 '21427700000000 ' 88.2
 101.8 78.5 9.6 36.6 328239523.0 37.09024 -95.712891]


In [ ]:
print(X_SelfVsInherited_Classification[0,:])

[180000 51.0 0 0 7 0 1 1971.0 6.0 28.0 117.24 7.5 '21427700000000 ' 88.2
 101.8 78.5 9.6 36.6 328239523.0 37.09024 -95.712891]


###Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_Wealth_train, X_Wealth_test, y_Wealth_train, y_Wealth_test = train_test_split(X_Wealth_Classification, y_Wealth_Classification, test_size = 0.4)

In [ ]:
X_Industry_train, X_Industry_test, y_Industry_train, y_Industry_test = train_test_split(X_Industry_Classification, y_Industry_Classification, test_size = 0.4)

In [ ]:
X_SelfVsInherited_train, X_SelfVsInherited_test, y_SelfVsInherited_train, y_SelfVsInherited_test = train_test_split(X_SelfVsInherited_Classification, y_SelfVsInherited_Classification, test_size = 0.4)

###Feature Scaling


In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_Wealth_train[:, :] = sc.fit_transform(X_Wealth_train[:, :])
X_Wealth_test[:, :] = sc.transform(X_Wealth_test[:, :])

X_Industry_train[:, :] = sc.fit_transform(X_Industry_train[:, :])
X_Industry_test[:, :] = sc.transform(X_Industry_test[:, :])

X_SelfVsInherited_train[:, :] = sc.fit_transform(X_SelfVsInherited_train[:, :])
X_SelfVsInherited_test[:, :] = sc.transform(X_SelfVsInherited_test[:, :])

In [ ]:
print(X_Wealth_train)

[[-1.9212737658582486 0.0 0.9834050368142934 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [1.4392649125836048 0.0 -1.4772736456108988 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.34387805965084806 0.0 -0.8621039750046008 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 ...
 [0.6162758484753958 0.0 0.36823536620799535 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [1.5078473345926222 0.0 0.5732919230767614 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.82395501371397 0.0 1.1884615936830594 ... 0.0 -9.237055564881302e-14
  -2.5579538487363607e-13]]


In [ ]:
print(X_Industry_train)

[[-0.1905364852139936 0.8703811457323521 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.4339484933143619 -0.35459972603910644 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [0.7597065464085979 0.6124904358857293 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 ...
 [4.326628665110149 0.03223633873082786 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.42458649300280926 0.8059084682706964 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.3918194919123751 0.225654371115795 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]]


In [ ]:
print(X_SelfVsInherited_train)

[[-0.38313496891824117 -0.07382354720808756 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.16829550113740663 -1.6462651027403448 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.25813746039121016 -0.07382354720808756 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 ...
 [-0.3323547310791348 -0.008305149060910176 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [-0.33626090322060453 -1.187636315710103 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]
 [0.20279085230221672 1.6951732027657018 0.0 ... 0.0
  -9.237055564881302e-14 -2.5579538487363607e-13]]


##  Baseline Classifier Implementation



### Nearest Neighbor Classifier For Wealth Classification

####Training the model (with 3 nearest neighbors each time)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN_3 = KNeighborsClassifier(n_neighbors = 3, metric = 'minkowski')
classifierKNN_3.fit(X_Wealth_train, y_Wealth_train)

KNeighborsClassifier(n_neighbors=3)

#####Predicting the test Results

In [ ]:
y_pred = classifierKNN_3.predict(X_Wealth_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_Wealth_test.reshape(len(y_Wealth_test),1)),1))

[[2 2]
 [2 2]
 [2 1]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 1]
 [2 0]
 [0 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [0 2]
 [2 2]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [1 2]
 [2 2]
 [2 2]
 [2 1]]


#####Making the Confusion Matrix

*accuracy for the prediction of Wealth Class*

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_Wealth_test, y_pred)
print(cm)
accuracy_score(y_Wealth_test, y_pred)

[[ 0  1  3]
 [ 1  0 11]
 [ 3  4 73]]


0.7604166666666666

####Training the model (with 1 nearest neighbors each time)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN_1 = KNeighborsClassifier(n_neighbors = 1, metric = 'minkowski', p = 2)
classifierKNN_1.fit(X_Wealth_train, y_Wealth_train)

KNeighborsClassifier(n_neighbors=1)

#####Predicting the test Results

In [ ]:
y_pred = classifierKNN_1.predict(X_Wealth_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_Wealth_test.reshape(len(y_Wealth_test),1)),1))

[[1 2]
 [2 2]
 [2 1]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [1 2]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [1 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [2 2]
 [2 2]
 [1 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [1 1]
 [2 0]
 [1 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [1 1]
 [1 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 1]
 [2 2]
 [0 2]
 [2 2]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [1 1]]


#####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_Wealth_test, y_pred)
print(cm)
accuracy_score(y_Wealth_test, y_pred)

[[ 0  1  3]
 [ 0  4  8]
 [ 4 10 66]]


0.7291666666666666

### Nearest Neighbor Classifier For Industry Classification

####Training the model (with 3 nearest neighbors each time)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN_3_Industry = KNeighborsClassifier(n_neighbors = 1, metric = 'minkowski', p = 2)
classifierKNN_3_Industry.fit(X_Industry_train, y_Industry_train)

KNeighborsClassifier(n_neighbors=1)

#####Predicting the test Results

In [ ]:
y_pred = classifierKNN_3_Industry.predict(X_Industry_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_Industry_test.reshape(len(y_Industry_test),1)),1))

[[ 4  5]
 [ 4  4]
 [ 5  5]
 [15 11]
 [ 4  5]
 [ 5 15]
 [ 5  3]
 [ 5 15]
 [15 15]
 [15  6]
 [14 14]
 [15 15]
 [15  7]
 [ 5 15]
 [ 5 11]
 [ 5  5]
 [ 5  5]
 [ 5 15]
 [ 5  5]
 [12  4]
 [11 15]
 [ 4 15]
 [15 11]
 [15 15]
 [14  5]
 [ 4 15]
 [ 5  4]
 [ 5  5]
 [15  5]
 [14  6]
 [ 3  6]
 [ 8  6]
 [ 5  5]
 [ 4 12]
 [ 5  5]
 [ 5  6]
 [15  4]
 [ 8  5]
 [15  3]
 [ 8 15]
 [15 15]
 [15 15]
 [15 15]
 [ 4  5]
 [ 3 15]
 [15  5]
 [15  5]
 [ 5 15]
 [15  5]
 [15 15]
 [13 11]
 [15  4]
 [ 8 11]
 [ 5  5]
 [15  5]
 [11 15]
 [ 8  4]
 [ 5  5]
 [15 15]
 [ 5  5]
 [ 5 15]
 [15  5]
 [15 10]
 [ 5  5]
 [ 5  4]
 [ 5  5]
 [15  3]
 [15 15]
 [14 15]
 [11  0]
 [15 15]
 [15 10]
 [ 8 14]
 [ 3 12]
 [ 8 15]
 [15 15]
 [ 8  5]
 [15 15]
 [15  0]
 [ 5 11]
 [11 15]
 [ 4  7]
 [ 5  6]
 [15 15]
 [ 5  5]
 [15  6]
 [ 5  5]
 [ 5  5]
 [15 15]
 [15 15]
 [ 5  5]
 [11  5]
 [12 11]
 [15  5]
 [14  5]
 [ 8 15]]


#####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_Industry_test, y_pred)
print(cm)
accuracy_score(y_Industry_test, y_pred)

[[ 0  0  0  0  0  0  0  0  1  0  0  0  1]
 [ 0  0  0  1  0  0  0  0  0  0  0  0  2]
 [ 0  0  1  2  0  0  1  0  0  1  0  0  2]
 [ 0  0  3 16  0  0  2  0  1  0  0  2  7]
 [ 0  1  0  2  0  0  1  0  0  0  0  1  2]
 [ 0  0  1  0  0  0  0  0  0  0  0  0  1]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  2]
 [ 0  0  0  2  0  0  1  0  0  1  1  0  2]
 [ 0  1  1  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  1  0  0  0  0  1  0]
 [ 0  1  2  6  0  0  3  0  3  0  0  1 15]]


0.34375

####Training the model (with 1 nearest neighbors each time)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN_1 = KNeighborsClassifier(n_neighbors = 1, metric = 'minkowski', p = 2)
classifierKNN_1.fit(X_Industry_train, y_Industry_train)

KNeighborsClassifier(n_neighbors=1)

#####Predicting the test Results

In [ ]:
y_pred = classifierKNN_1.predict(X_Industry_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_Industry_test.reshape(len(y_Industry_test),1)),1))

[[ 4  5]
 [ 4  4]
 [ 5  5]
 [15 11]
 [ 4  5]
 [ 5 15]
 [ 5  3]
 [ 5 15]
 [15 15]
 [15  6]
 [14 14]
 [15 15]
 [15  7]
 [ 5 15]
 [ 5 11]
 [ 5  5]
 [ 5  5]
 [ 5 15]
 [ 5  5]
 [12  4]
 [11 15]
 [ 4 15]
 [15 11]
 [15 15]
 [14  5]
 [ 4 15]
 [ 5  4]
 [ 5  5]
 [15  5]
 [14  6]
 [ 3  6]
 [ 8  6]
 [ 5  5]
 [ 4 12]
 [ 5  5]
 [ 5  6]
 [15  4]
 [ 8  5]
 [15  3]
 [ 8 15]
 [15 15]
 [15 15]
 [15 15]
 [ 4  5]
 [ 3 15]
 [15  5]
 [15  5]
 [ 5 15]
 [15  5]
 [15 15]
 [13 11]
 [15  4]
 [ 8 11]
 [ 5  5]
 [15  5]
 [11 15]
 [ 8  4]
 [ 5  5]
 [15 15]
 [ 5  5]
 [ 5 15]
 [15  5]
 [15 10]
 [ 5  5]
 [ 5  4]
 [ 5  5]
 [15  3]
 [15 15]
 [14 15]
 [11  0]
 [15 15]
 [15 10]
 [ 8 14]
 [ 3 12]
 [ 8 15]
 [15 15]
 [ 8  5]
 [15 15]
 [15  0]
 [ 5 11]
 [11 15]
 [ 4  7]
 [ 5  6]
 [15 15]
 [ 5  5]
 [15  6]
 [ 5  5]
 [ 5  5]
 [15 15]
 [15 15]
 [ 5  5]
 [11  5]
 [12 11]
 [15  5]
 [14  5]
 [ 8 15]]


#####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_Industry_test, y_pred)
print(cm)
accuracy_score(y_Industry_test, y_pred)

[[ 0  0  0  0  0  0  0  0  1  0  0  0  1]
 [ 0  0  0  1  0  0  0  0  0  0  0  0  2]
 [ 0  0  1  2  0  0  1  0  0  1  0  0  2]
 [ 0  0  3 16  0  0  2  0  1  0  0  2  7]
 [ 0  1  0  2  0  0  1  0  0  0  0  1  2]
 [ 0  0  1  0  0  0  0  0  0  0  0  0  1]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  2]
 [ 0  0  0  2  0  0  1  0  0  1  1  0  2]
 [ 0  1  1  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  1  0  0  0  0  1  0]
 [ 0  1  2  6  0  0  3  0  3  0  0  1 15]]


0.34375

### Nearest Neighbor Classifier For Self-Made vs. Inherited Wealth

####Training the model (with 3 nearest neighbors each time)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN_3 = KNeighborsClassifier(n_neighbors = 1, metric = 'minkowski', p = 2)
classifierKNN_3.fit(X_SelfVsInherited_train, y_SelfVsInherited_train)

KNeighborsClassifier(n_neighbors=1)

#####Predicting the test Results

In [ ]:
y_pred = classifierKNN_3.predict(X_SelfVsInherited_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_SelfVsInherited_test.reshape(len(y_SelfVsInherited_test),1)),1))

[[1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [1 1]]


#####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_SelfVsInherited_test, y_pred)
print(cm)
accuracy_score(y_SelfVsInherited_test, y_pred)

[[ 8  9]
 [ 7 72]]


0.8333333333333334

####Training the model (with 1 nearest neighbors each time)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN_1 = KNeighborsClassifier(n_neighbors = 1, metric = 'minkowski', p = 2)
classifierKNN_1.fit(X_SelfVsInherited_train, y_SelfVsInherited_train)

KNeighborsClassifier(n_neighbors=1)

#####Predicting the test Results

In [ ]:
y_pred = classifierKNN_1.predict(X_SelfVsInherited_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_SelfVsInherited_test.reshape(len(y_SelfVsInherited_test),1)),1))

[[1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [1 1]]


#####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_SelfVsInherited_test, y_pred)
print(cm)
accuracy_score(y_SelfVsInherited_test, y_pred)

[[ 8  9]
 [ 7 72]]


0.8333333333333334

### Nearest Class Centroid Classifier For Wealth Classification


####Training the model

In [ ]:
from sklearn.neighbors import NearestCentroid
classifierNC = NearestCentroid(metric='euclidean')
classifierNC.fit(X_Wealth_train, y_Wealth_train)

NearestCentroid()

####Predicting the test Results

In [ ]:
y_pred = classifierNC.predict(X_Wealth_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_Wealth_test.reshape(len(y_Wealth_test),1)),1))

[[1 2]
 [2 2]
 [0 1]
 [1 1]
 [1 2]
 [1 2]
 [1 2]
 [0 2]
 [1 2]
 [1 2]
 [1 2]
 [1 1]
 [2 2]
 [1 2]
 [2 0]
 [1 2]
 [2 2]
 [2 2]
 [1 2]
 [1 2]
 [2 2]
 [2 2]
 [2 2]
 [1 1]
 [2 2]
 [0 2]
 [2 2]
 [0 2]
 [0 2]
 [1 2]
 [0 2]
 [0 2]
 [1 2]
 [1 0]
 [1 2]
 [2 2]
 [1 2]
 [2 2]
 [1 2]
 [1 2]
 [2 2]
 [2 1]
 [1 2]
 [2 2]
 [1 1]
 [0 0]
 [0 1]
 [1 2]
 [1 2]
 [2 2]
 [1 2]
 [1 1]
 [1 2]
 [1 2]
 [2 2]
 [1 2]
 [1 1]
 [1 2]
 [2 1]
 [1 2]
 [2 2]
 [1 2]
 [0 2]
 [0 2]
 [1 2]
 [0 2]
 [1 2]
 [2 2]
 [1 2]
 [1 2]
 [2 2]
 [0 2]
 [1 2]
 [1 2]
 [0 2]
 [1 1]
 [2 2]
 [1 2]
 [2 2]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [0 2]
 [1 2]
 [0 2]
 [1 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [2 2]
 [1 2]
 [2 2]
 [1 1]]


####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_Wealth_test, y_pred)
print(cm)
accuracy_score(y_Wealth_test, y_pred)

[[ 2  1  1]
 [ 2  8  2]
 [15 37 28]]


0.3958333333333333

### Nearest Class Centroid Classifier For Industry Classification


####Training the model

In [ ]:
from sklearn.neighbors import NearestCentroid
classifierNC = NearestCentroid(metric='euclidean')
classifierNC.fit(X_Industry_train, y_Industry_train)

NearestCentroid()

####Predicting the test Results

In [ ]:
y_pred = classifierNC.predict(X_Industry_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_Industry_test.reshape(len(y_Industry_test),1)),1))

[[ 9  5]
 [ 1  4]
 [ 5  5]
 [ 3 11]
 [ 4  5]
 [15 15]
 [ 5  3]
 [15 15]
 [ 6 15]
 [ 3  6]
 [ 8 14]
 [15 15]
 [ 8  7]
 [15 15]
 [ 8 11]
 [15  5]
 [11  5]
 [ 6 15]
 [ 5  5]
 [12  4]
 [ 4 15]
 [ 6 15]
 [ 3 11]
 [ 1 15]
 [ 6  5]
 [ 1 15]
 [13  4]
 [ 9  5]
 [ 5  5]
 [ 3  6]
 [ 3  6]
 [15  6]
 [ 9  5]
 [12 12]
 [15  5]
 [ 6  6]
 [ 1  4]
 [ 8  5]
 [ 3  3]
 [ 8 15]
 [15 15]
 [15 15]
 [15 15]
 [ 4  5]
 [14 15]
 [ 8  5]
 [ 6  5]
 [ 7 15]
 [15  5]
 [15 15]
 [13 11]
 [ 2  4]
 [ 8 11]
 [ 5  5]
 [ 6  5]
 [15 15]
 [ 8  4]
 [12  5]
 [ 8 15]
 [ 9  5]
 [15 15]
 [ 8  5]
 [ 7 10]
 [ 8  5]
 [14  4]
 [ 5  5]
 [ 6  3]
 [15 15]
 [ 6 15]
 [14  0]
 [ 1 15]
 [ 9 10]
 [ 6 14]
 [ 3 12]
 [ 5 15]
 [15 15]
 [ 6  5]
 [15 15]
 [ 2  0]
 [12 11]
 [ 9 15]
 [ 1  7]
 [ 4  6]
 [15 15]
 [ 7  5]
 [ 9  6]
 [14  5]
 [ 5  5]
 [ 8 15]
 [15 15]
 [ 5  5]
 [ 9  5]
 [12 11]
 [ 5  5]
 [ 5  5]
 [ 8 15]]


####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_Industry_test, y_pred)
print(cm)
accuracy_score(y_Industry_test, y_pred)

[[ 0  0  1  0  0  0  0  0  0  0  0  0  0  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  1  0  1  1  0  0  0  0  0  0  0  0  0]
 [ 0  2  1  0  0  0  0  0  1  0  0  0  1  1  1  0]
 [ 0  0  0  0  2  9  4  1  4  5  0  1  1  0  1  3]
 [ 0  0  0  3  1  0  1  0  0  1  0  0  0  0  0  1]
 [ 0  1  0  0  0  0  0  0  1  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  1  0  1  0  0  0  0  0  0]
 [ 0  0  0  2  0  0  0  0  2  0  0  0  2  1  0  0]
 [ 0  0  0  1  0  0  0  0  0  0  0  0  1  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  1  0  1  0  0  0  0  0  0  0]
 [ 0  3  0  0  1  1  4  1  4  1  0  0  0  0  1 15]]


0.28125

### Nearest Class Centroid Classifier For Self-Made vs. Inherited Wealth


####Training the model

In [ ]:
from sklearn.neighbors import NearestCentroid
classifierNC = NearestCentroid(metric='euclidean')
classifierNC.fit(X_SelfVsInherited_train, y_SelfVsInherited_train)

NearestCentroid()

####Predicting the test Results

In [ ]:
y_pred = classifierNC.predict(X_SelfVsInherited_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_SelfVsInherited_test.reshape(len(y_SelfVsInherited_test),1)),1))

[[1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [0 1]
 [0 1]
 [0 1]
 [1 1]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [1 1]
 [0 1]
 [0 1]
 [0 1]
 [0 0]
 [1 1]
 [1 1]]


####Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_SelfVsInherited_test, y_pred)
print(cm)
accuracy_score(y_SelfVsInherited_test, y_pred)

[[13  4]
 [27 52]]


0.6770833333333334

##Neural Network Model Implementation


###For Wealth Classification

In [ ]:
from tensorflow.keras.utils import to_categorical

y_Wealth_train_one_hot = to_categorical(y_Wealth_train)
y_Wealth_test_one_hot = to_categorical(y_Wealth_test)

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

####NN First Attempt

In [ ]:
ann1 = tf.keras.models.Sequential()
ann1.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann1.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann1.add(tf.keras.layers.Dense(units=3, activation='sigmoid'))

In [ ]:
from tensorflow.keras.utils import to_categorical

y_Wealth_train_one_hot = to_categorical(y_Wealth_train)
y_Wealth_test_one_hot = to_categorical(y_Wealth_test)

In [ ]:

# Compile the model with the correct loss function for multi-class classification
ann1.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann1.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 1.2090 - accuracy: 0.2113
Epoch 2/100
5/5 [==============================] - 0s 4ms/step - loss: 1.1803 - accuracy: 0.2606
Epoch 3/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1531 - accuracy: 0.2817
Epoch 4/100
5/5 [==============================] - 0s 4ms/step - loss: 1.1279 - accuracy: 0.3451
Epoch 5/100
5/5 [==============================] - 0s 4ms/step - loss: 1.1036 - accuracy: 0.3662
Epoch 6/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0800 - accuracy: 0.4366
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0573 - accuracy: 0.5423
Epoch 8/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0326 - accuracy: 0.5775
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0112 - accuracy: 0.5986
Epoch 10/100
5/5 [==============================] - 0s 3ms/step - loss: 0.9884 - accuracy: 0.6549
Epoch 11/100
5/5 [===========

In [ ]:
ann1.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 5ms/step - loss: 0.5525 - accuracy: 0.8333


[0.5524637699127197, 0.8333333134651184]

####NN Second Attempt

based on the classification I made we see these:
Ultra Wealth (Category 1): 10 instances (4.20%)
Median Wealth (Category 2): 32 instances (13.45%)
Low Wealth (Category 3): 196 instances (82.35%)

Since this distribution indicates that my classes are indeed imbalanced. The majority of data falls into the "Low Wealth" category.

To address this imbalance, I will consider the following approach:

To create different weight for the different classes

In [ ]:
from sklearn.utils.class_weight import compute_class_weight

# Calculate class weights
class_weights = compute_class_weight('balanced',
                                     classes=np.unique(y_Wealth_train),
                                     y=y_Wealth_train)
class_weights = dict(zip(np.unique(y_Wealth_train), class_weights))


In [ ]:
# Build the neural network
ann2 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=6, activation='relu'),
    tf.keras.layers.Dense(units=6, activation='relu'),
    tf.keras.layers.Dense(units=3, activation='sigmoid')  # Assuming 4 output classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann2.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 1.1038 - accuracy: 0.2817
Epoch 2/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0982 - accuracy: 0.3099
Epoch 3/100
5/5 [==============================] - 0s 5ms/step - loss: 1.0932 - accuracy: 0.3169
Epoch 4/100
5/5 [==============================] - 0s 5ms/step - loss: 1.0888 - accuracy: 0.3239
Epoch 5/100
5/5 [==============================] - 0s 5ms/step - loss: 1.0845 - accuracy: 0.3310
Epoch 6/100
5/5 [==============================] - 0s 6ms/step - loss: 1.0806 - accuracy: 0.3310
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0765 - accuracy: 0.3451
Epoch 8/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0731 - accuracy: 0.3592
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0689 - accuracy: 0.3592
Epoch 10/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0651 - accuracy: 0.3662
Epoch 11/100
5/5 [===========

In [ ]:
ann2.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 6ms/step - loss: 1.1035 - accuracy: 0.4167


[1.1035469770431519, 0.4166666567325592]

Since the previous accuracy was Close the the Percetange of the Low Income Billioners I will assume that it was wrong and the model didn't predict the other ones

####NN Third Attempt

I will try different output function

In [ ]:
# Build the neural network
ann3 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=6, activation='relu'),
    tf.keras.layers.Dense(units=6, activation='relu'),
    tf.keras.layers.Dense(units=3, activation='softmax')  # Assuming 4 output classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann3.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann3.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 1.1842 - accuracy: 0.1479
Epoch 2/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1735 - accuracy: 0.1408
Epoch 3/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1652 - accuracy: 0.1479
Epoch 4/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1573 - accuracy: 0.1549
Epoch 5/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1498 - accuracy: 0.1690
Epoch 6/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1423 - accuracy: 0.1620
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1361 - accuracy: 0.1761
Epoch 8/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1307 - accuracy: 0.1972
Epoch 9/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1246 - accuracy: 0.2042
Epoch 10/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1199 - accuracy: 0.2042
Epoch 11/100
5/5 [===========

In [ ]:
ann3.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 5ms/step - loss: 1.0624 - accuracy: 0.4167


[1.0623700618743896, 0.4166666567325592]

####NN Fourth Attempt

I will double the neurons in hidden layers

In [ ]:
ann4 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=16, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=12, activation='relu'),  # Slightly fewer neurons in the second layer
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann4.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann4.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 1.2172 - accuracy: 0.5634
Epoch 2/100
5/5 [==============================] - 0s 4ms/step - loss: 1.1940 - accuracy: 0.5775
Epoch 3/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1735 - accuracy: 0.6056
Epoch 4/100
5/5 [==============================] - 0s 5ms/step - loss: 1.1587 - accuracy: 0.5986
Epoch 5/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1441 - accuracy: 0.6127
Epoch 6/100
5/5 [==============================] - 0s 4ms/step - loss: 1.1310 - accuracy: 0.6338
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1183 - accuracy: 0.6479
Epoch 8/100
5/5 [==============================] - 0s 4ms/step - loss: 1.1092 - accuracy: 0.6620
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0971 - accuracy: 0.6690
Epoch 10/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0895 - accuracy: 0.6690
Epoch 11/100
5/5 [===========

In [ ]:
ann4.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 6ms/step - loss: 0.8497 - accuracy: 0.5625


[0.8497211337089539, 0.5625]

Indeed improved accuracy

####NN Fifth Attempt

I will double the neurons in hidden layers once more

In [ ]:
ann5 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=32, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=24, activation='relu'),  # Slightly fewer neurons in the second layer
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann5.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann5.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 1.1298 - accuracy: 0.3521
Epoch 2/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0936 - accuracy: 0.4014
Epoch 3/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0635 - accuracy: 0.4296
Epoch 4/100
5/5 [==============================] - 0s 4ms/step - loss: 1.0405 - accuracy: 0.4366
Epoch 5/100
5/5 [==============================] - 0s 6ms/step - loss: 1.0195 - accuracy: 0.4648
Epoch 6/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9968 - accuracy: 0.4859
Epoch 7/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9782 - accuracy: 0.5000
Epoch 8/100
5/5 [==============================] - 0s 5ms/step - loss: 0.9636 - accuracy: 0.5282
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9443 - accuracy: 0.5493
Epoch 10/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9315 - accuracy: 0.5563
Epoch 11/100
5/5 [===========

In [ ]:
ann5.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 5ms/step - loss: 0.8153 - accuracy: 0.6354


[0.8153050541877747, 0.6354166865348816]

Indeed improved accuracy

####NN Sixth Attempt

I will double the neurons in hidden layers once more

In [ ]:
ann6 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=64, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=48, activation='relu'),  # Slightly fewer neurons in the second layer
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann6.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann6.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 5ms/step - loss: 1.0965 - accuracy: 0.6549
Epoch 2/100
5/5 [==============================] - 0s 6ms/step - loss: 1.0358 - accuracy: 0.6338
Epoch 3/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9996 - accuracy: 0.5845
Epoch 4/100
5/5 [==============================] - 0s 6ms/step - loss: 0.9640 - accuracy: 0.5563
Epoch 5/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9358 - accuracy: 0.5775
Epoch 6/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9087 - accuracy: 0.5563
Epoch 7/100
5/5 [==============================] - 0s 4ms/step - loss: 0.8829 - accuracy: 0.5141
Epoch 8/100
5/5 [==============================] - 0s 5ms/step - loss: 0.8625 - accuracy: 0.5282
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 0.8383 - accuracy: 0.5141
Epoch 10/100
5/5 [==============================] - 0s 4ms/step - loss: 0.8183 - accuracy: 0.5141
Epoch 11/100
5/5 [===========

In [ ]:
ann6.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 5ms/step - loss: 0.9136 - accuracy: 0.6771


[0.9135701060295105, 0.6770833134651184]

Indeed improved accuracy

####NN Seventh Attempt

I will double the neurons in hidden layers once more and Increce epochs cause of the complexity

In [ ]:
ann7 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=138, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=96, activation='relu'),
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann7.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann7.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 6ms/step - loss: 1.1019 - accuracy: 0.2606
Epoch 2/100
5/5 [==============================] - 0s 5ms/step - loss: 0.9849 - accuracy: 0.3944
Epoch 3/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9097 - accuracy: 0.4648
Epoch 4/100
5/5 [==============================] - 0s 4ms/step - loss: 0.8495 - accuracy: 0.5070
Epoch 5/100
5/5 [==============================] - 0s 4ms/step - loss: 0.8061 - accuracy: 0.5423
Epoch 6/100
5/5 [==============================] - 0s 3ms/step - loss: 0.7607 - accuracy: 0.5775
Epoch 7/100
5/5 [==============================] - 0s 4ms/step - loss: 0.7296 - accuracy: 0.5986
Epoch 8/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6983 - accuracy: 0.6338
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6739 - accuracy: 0.6620
Epoch 10/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6522 - accuracy: 0.6338
Epoch 11/100
5/5 [===========

In [ ]:
ann7.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 6ms/step - loss: 1.1777 - accuracy: 0.6562


[1.1776841878890991, 0.65625]

Indeed improved accuracy



####NN Eightth Attempt

I will triple the epochs to help train in the bigger complexity

In [ ]:
ann8 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=138, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=96, activation='relu'),
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann8.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann8.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=300, class_weight=class_weights)

Epoch 1/300
5/5 [==============================] - 1s 4ms/step - loss: 1.0822 - accuracy: 0.6549
Epoch 2/300
5/5 [==============================] - 0s 4ms/step - loss: 0.9828 - accuracy: 0.5563
Epoch 3/300
5/5 [==============================] - 0s 4ms/step - loss: 0.9005 - accuracy: 0.5141
Epoch 4/300
5/5 [==============================] - 0s 3ms/step - loss: 0.8589 - accuracy: 0.5282
Epoch 5/300
5/5 [==============================] - 0s 4ms/step - loss: 0.8158 - accuracy: 0.5282
Epoch 6/300
5/5 [==============================] - 0s 4ms/step - loss: 0.7778 - accuracy: 0.5211
Epoch 7/300
5/5 [==============================] - 0s 3ms/step - loss: 0.7447 - accuracy: 0.5282
Epoch 8/300
5/5 [==============================] - 0s 4ms/step - loss: 0.7165 - accuracy: 0.5423
Epoch 9/300
5/5 [==============================] - 0s 4ms/step - loss: 0.6925 - accuracy: 0.5775
Epoch 10/300
5/5 [==============================] - 0s 4ms/step - loss: 0.6719 - accuracy: 0.6056
Epoch 11/300
5/5 [===========

In [ ]:
ann8.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 7ms/step - loss: 2.4345 - accuracy: 0.7292


[2.4344534873962402, 0.7291666865348816]

Although I reached accuracy one on the train set since the accuracy on the test set improved I will accume i don't have overfitting

####NN Ninght Attempt

Trying a different optimizer

In [ ]:
ann9 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=138, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=96, activation='relu'),
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann9.compile(optimizer='RMSprop', loss='categorical_crossentropy', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann9.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=300, class_weight=class_weights)

Epoch 1/300
5/5 [==============================] - 1s 5ms/step - loss: 1.0859 - accuracy: 0.3732
Epoch 2/300
5/5 [==============================] - 0s 6ms/step - loss: 0.9579 - accuracy: 0.4859
Epoch 3/300
5/5 [==============================] - 0s 6ms/step - loss: 0.8881 - accuracy: 0.4859
Epoch 4/300
5/5 [==============================] - 0s 4ms/step - loss: 0.8495 - accuracy: 0.5000
Epoch 5/300
5/5 [==============================] - 0s 5ms/step - loss: 0.8094 - accuracy: 0.4930
Epoch 6/300
5/5 [==============================] - 0s 4ms/step - loss: 0.7762 - accuracy: 0.5141
Epoch 7/300
5/5 [==============================] - 0s 5ms/step - loss: 0.7493 - accuracy: 0.5211
Epoch 8/300
5/5 [==============================] - 0s 4ms/step - loss: 0.7250 - accuracy: 0.5563
Epoch 9/300
5/5 [==============================] - 0s 6ms/step - loss: 0.7008 - accuracy: 0.5563
Epoch 10/300
5/5 [==============================] - 0s 6ms/step - loss: 0.6901 - accuracy: 0.5845
Epoch 11/300
5/5 [===========

In [ ]:
ann9.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 5ms/step - loss: 3.0091 - accuracy: 0.7500


[3.0091464519500732, 0.75]

Trying a different Loss function

In [ ]:
ann9 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=138, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=96, activation='relu'),
    tf.keras.layers.Dense(units=3, activation='softmax') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann9.compile(optimizer='adam', loss='CategoricalHinge', metrics=['accuracy'])

X_Wealth_train = np.asarray(X_Wealth_train).astype('float32')
X_Wealth_test = np.asarray(X_Wealth_test).astype('float32')

ann9.fit(X_Wealth_train, y_Wealth_train_one_hot, batch_size=32, epochs=300, class_weight=class_weights)

Epoch 1/300
5/5 [==============================] - 1s 8ms/step - loss: 1.0162 - accuracy: 0.4648
Epoch 2/300
5/5 [==============================] - 0s 5ms/step - loss: 0.9719 - accuracy: 0.5352
Epoch 3/300
5/5 [==============================] - 0s 4ms/step - loss: 0.9411 - accuracy: 0.4859
Epoch 4/300
5/5 [==============================] - 0s 3ms/step - loss: 0.9166 - accuracy: 0.4648
Epoch 5/300
5/5 [==============================] - 0s 4ms/step - loss: 0.8865 - accuracy: 0.4155
Epoch 6/300
5/5 [==============================] - 0s 5ms/step - loss: 0.8541 - accuracy: 0.4085
Epoch 7/300
5/5 [==============================] - 0s 4ms/step - loss: 0.8293 - accuracy: 0.4155
Epoch 8/300
5/5 [==============================] - 0s 5ms/step - loss: 0.8018 - accuracy: 0.4155
Epoch 9/300
5/5 [==============================] - 0s 4ms/step - loss: 0.7769 - accuracy: 0.4225
Epoch 10/300
5/5 [==============================] - 0s 4ms/step - loss: 0.7525 - accuracy: 0.4437
Epoch 11/300
5/5 [===========

In [ ]:
ann9.evaluate(X_Wealth_test,y_Wealth_test_one_hot)

3/3 [==============================] - 0s 6ms/step - loss: 0.6054 - accuracy: 0.7083


[0.6053510308265686, 0.7083333134651184]

worst results on both

###For Industry Classification

Based on the bad results I had previously I will not make this accuming thering a good ammount of features to predict this

###For Self-Made vs. Inherited Wealth Classification

In [ ]:
X_SelfVsInherited_train = np.asarray(X_SelfVsInherited_train).astype('float32')
X_SelfVsInherited_test = np.asarray(X_SelfVsInherited_test).astype('float32')

In [ ]:
print(y_SelfVsInherited_train)

[1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0 1 1 0 1 1 0 1 0 1 1 1 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 0 1 1 1 1 0 1 1 1 0 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 0 1 1 1 0]


####NN First Attempt

I will be using a binary_crossentropy loss function since I have binary classification

In [ ]:
ann1 = tf.keras.models.Sequential()
ann1.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann1.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann1.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann1.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann1.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=32, epochs=100)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 0.7198 - accuracy: 0.4155
Epoch 2/100
5/5 [==============================] - 0s 3ms/step - loss: 0.7079 - accuracy: 0.4648
Epoch 3/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6971 - accuracy: 0.5000
Epoch 4/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6861 - accuracy: 0.5493
Epoch 5/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6755 - accuracy: 0.5704
Epoch 6/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6646 - accuracy: 0.5986
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6539 - accuracy: 0.6197
Epoch 8/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6429 - accuracy: 0.7254
Epoch 9/100
5/5 [==============================] - 0s 5ms/step - loss: 0.6317 - accuracy: 0.7606
Epoch 10/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6201 - accuracy: 0.7887
Epoch 11/100
5/5 [===========

In [ ]:
ann1.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 0.3807 - accuracy: 0.8333


[0.3806510865688324, 0.8333333134651184]

Since Here I have a binary classification I will try to print the predictions allong side the actuall values

In [ ]:
y_pred = ann1.predict(X_SelfVsInherited_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_SelfVsInherited_test.reshape(len(y_SelfVsInherited_test),1)),1))

3/3 [==============================] - 0s 4ms/step
[[1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]]


####NN Second Attempt

In [ ]:
y_array = np.array(y_SelfVsInherited_train)

# Calculate percentages
percent_self_made = np.mean(y_array) * 100
percent_not_self_made = 100 - percent_self_made

# Print the percentages
print(f"Percentage of Self-Made Billionaires: {percent_self_made:.2f}%")
print(f"Percentage of Not Self-Made Billionaires: {percent_not_self_made:.2f}%")

Percentage of Self-Made Billionaires: 83.80%
Percentage of Not Self-Made Billionaires: 16.20%


This indicated that I have inbalanced dataset once more. So I will try the previous method with different weights

In [ ]:
from sklearn.utils.class_weight import compute_class_weight

# Calculate class weights
class_weights = compute_class_weight('balanced',
                                     classes=np.unique(y_SelfVsInherited_train),
                                     y=y_SelfVsInherited_train)
class_weights = dict(zip(np.unique(y_SelfVsInherited_train), class_weights))

In [ ]:
ann2 = tf.keras.models.Sequential()
ann2.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann2.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann2.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann2.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann2.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 2s 4ms/step - loss: 0.7370 - accuracy: 0.1761
Epoch 2/100
5/5 [==============================] - 0s 4ms/step - loss: 0.7293 - accuracy: 0.1761
Epoch 3/100
5/5 [==============================] - 0s 3ms/step - loss: 0.7222 - accuracy: 0.1761
Epoch 4/100
5/5 [==============================] - 0s 3ms/step - loss: 0.7152 - accuracy: 0.1831
Epoch 5/100
5/5 [==============================] - 0s 3ms/step - loss: 0.7095 - accuracy: 0.1901
Epoch 6/100
5/5 [==============================] - 0s 4ms/step - loss: 0.7047 - accuracy: 0.1972
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6996 - accuracy: 0.2183
Epoch 8/100
5/5 [==============================] - 0s 5ms/step - loss: 0.6951 - accuracy: 0.2324
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6918 - accuracy: 0.2465
Epoch 10/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6876 - accuracy: 0.2535
Epoch 11/100
5/5 [===========

In [ ]:
ann2.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 0.7213 - accuracy: 0.6562


[0.7213227152824402, 0.65625]

Since Here I have a binary classification I will try to print the predictions allong side the actuall values

In [ ]:
y_pred = ann1.predict(X_SelfVsInherited_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_SelfVsInherited_test.reshape(len(y_SelfVsInherited_test),1)),1))

3/3 [==============================] - 0s 4ms/step
[[1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]]


Here I have a worst accuracy but my model is train is the inbalanced dataset.

####NN Third Attempt

I will try to increase the complexity of my model

In [ ]:
ann3 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=32, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=12, activation='relu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann3.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann3.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=32, epochs=100, class_weight=class_weights)

Epoch 1/100
5/5 [==============================] - 1s 4ms/step - loss: 0.7056 - accuracy: 0.5000
Epoch 2/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6893 - accuracy: 0.5282
Epoch 3/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6760 - accuracy: 0.5563
Epoch 4/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6641 - accuracy: 0.6268
Epoch 5/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6546 - accuracy: 0.6408
Epoch 6/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6437 - accuracy: 0.6479
Epoch 7/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6343 - accuracy: 0.6620
Epoch 8/100
5/5 [==============================] - 0s 4ms/step - loss: 0.6249 - accuracy: 0.6761
Epoch 9/100
5/5 [==============================] - 0s 3ms/step - loss: 0.6157 - accuracy: 0.6831
Epoch 10/100
5/5 [==============================] - 0s 6ms/step - loss: 0.6081 - accuracy: 0.7042
Epoch 11/100
5/5 [===========

In [ ]:
ann3.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 7ms/step - loss: 0.5630 - accuracy: 0.7188


[0.5630003809928894, 0.71875]

Here I have improved the accuracy

####NN Fourth Attempt

I will try to have more epochs and a lower bach size for a better train

In [ ]:
ann4 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=32, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=12, activation='relu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann4.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann4.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=16, epochs=200, class_weight=class_weights)

Epoch 1/200
9/9 [==============================] - 1s 4ms/step - loss: 0.7168 - accuracy: 0.5845
Epoch 2/200
9/9 [==============================] - 0s 5ms/step - loss: 0.6758 - accuracy: 0.6408
Epoch 3/200
9/9 [==============================] - 0s 6ms/step - loss: 0.6442 - accuracy: 0.7324
Epoch 4/200
9/9 [==============================] - 0s 4ms/step - loss: 0.6229 - accuracy: 0.7746
Epoch 5/200
9/9 [==============================] - 0s 4ms/step - loss: 0.6000 - accuracy: 0.7606
Epoch 6/200
9/9 [==============================] - 0s 4ms/step - loss: 0.5820 - accuracy: 0.7606
Epoch 7/200
9/9 [==============================] - 0s 4ms/step - loss: 0.5659 - accuracy: 0.7254
Epoch 8/200
9/9 [==============================] - 0s 4ms/step - loss: 0.5523 - accuracy: 0.7324
Epoch 9/200
9/9 [==============================] - 0s 3ms/step - loss: 0.5418 - accuracy: 0.7254
Epoch 10/200
9/9 [==============================] - 0s 3ms/step - loss: 0.5298 - accuracy: 0.7254
Epoch 11/200
9/9 [===========

In [ ]:
ann4.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 0.7836 - accuracy: 0.7500


[0.7835862040519714, 0.75]

Here I have improved the accuracy

####NN Fifth Attempt

I will try to increase complexity and to have more epochs and a lower bach size for a better train again

In [ ]:
ann5 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=64, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=24, activation='relu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann5.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann5.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=16, epochs=300, class_weight=class_weights)

Epoch 1/300
9/9 [==============================] - 1s 4ms/step - loss: 0.7004 - accuracy: 0.7394
Epoch 2/300
9/9 [==============================] - 0s 3ms/step - loss: 0.6631 - accuracy: 0.7254
Epoch 3/300
9/9 [==============================] - 0s 3ms/step - loss: 0.6375 - accuracy: 0.7324
Epoch 4/300
9/9 [==============================] - 0s 3ms/step - loss: 0.6166 - accuracy: 0.7113
Epoch 5/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5967 - accuracy: 0.6972
Epoch 6/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5763 - accuracy: 0.7113
Epoch 7/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5586 - accuracy: 0.7113
Epoch 8/300
9/9 [==============================] - 0s 4ms/step - loss: 0.5425 - accuracy: 0.7113
Epoch 9/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5295 - accuracy: 0.7183
Epoch 10/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5153 - accuracy: 0.7113
Epoch 11/300
9/9 [===========

In [ ]:
ann5.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 4ms/step - loss: 1.8373 - accuracy: 0.8021


[1.8372774124145508, 0.8020833134651184]

Here I have improved the accuracy slitly

####NN Sixth Attempt

I will try to increase only complexity once more

In [ ]:
ann6 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=138, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=48, activation='relu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann6.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann6.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=16, epochs=300, class_weight=class_weights)

Epoch 1/300
9/9 [==============================] - 1s 3ms/step - loss: 0.6951 - accuracy: 0.5775
Epoch 2/300
9/9 [==============================] - 0s 4ms/step - loss: 0.6260 - accuracy: 0.6831
Epoch 3/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5825 - accuracy: 0.6901
Epoch 4/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5500 - accuracy: 0.7042
Epoch 5/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5255 - accuracy: 0.7394
Epoch 6/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5067 - accuracy: 0.7324
Epoch 7/300
9/9 [==============================] - 0s 3ms/step - loss: 0.4902 - accuracy: 0.7465
Epoch 8/300
9/9 [==============================] - 0s 2ms/step - loss: 0.4766 - accuracy: 0.7113
Epoch 9/300
9/9 [==============================] - 0s 2ms/step - loss: 0.4612 - accuracy: 0.7535
Epoch 10/300
9/9 [==============================] - 0s 3ms/step - loss: 0.4513 - accuracy: 0.7535
Epoch 11/300
9/9 [===========

In [ ]:
ann6.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 1.7199 - accuracy: 0.7917


[1.7198501825332642, 0.7916666865348816]

Accuracy remained the same so I will keep the previous simpler model

####NN Seventh Attempt

I will try to increase complexity by adding on more layer of neurons

In [ ]:
ann7 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=64, activation='relu'),  # Increased neurons
    tf.keras.layers.Dense(units=24, activation='relu'),
    tf.keras.layers.Dense(units=64, activation='relu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann7.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann7.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=16, epochs=300, class_weight=class_weights)

Epoch 1/300
9/9 [==============================] - 1s 5ms/step - loss: 0.7178 - accuracy: 0.5070
Epoch 2/300
9/9 [==============================] - 0s 4ms/step - loss: 0.6749 - accuracy: 0.5775
Epoch 3/300
9/9 [==============================] - 0s 4ms/step - loss: 0.6503 - accuracy: 0.6620
Epoch 4/300
9/9 [==============================] - 0s 3ms/step - loss: 0.6258 - accuracy: 0.6408
Epoch 5/300
9/9 [==============================] - 0s 4ms/step - loss: 0.6023 - accuracy: 0.6338
Epoch 6/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5801 - accuracy: 0.6338
Epoch 7/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5538 - accuracy: 0.6620
Epoch 8/300
9/9 [==============================] - 0s 4ms/step - loss: 0.5316 - accuracy: 0.6408
Epoch 9/300
9/9 [==============================] - 0s 4ms/step - loss: 0.5125 - accuracy: 0.6831
Epoch 10/300
9/9 [==============================] - 0s 4ms/step - loss: 0.4926 - accuracy: 0.7183
Epoch 11/300
9/9 [===========

In [ ]:
ann7.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 1.8710 - accuracy: 0.8229


[1.8710042238235474, 0.8229166865348816]

Worst accuracy maybe overfitted to the dataset or (the error didn't manage to reach back layers as much)

####NN Eight Attempt

I will try a different activation function in hidden layers

In [ ]:
ann8 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=64, activation='selu'),  # Increased neurons
    tf.keras.layers.Dense(units=24, activation='selu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann8.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann8.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=16, epochs=300, class_weight=class_weights)

Epoch 1/300
9/9 [==============================] - 1s 3ms/step - loss: 0.7218 - accuracy: 0.6338
Epoch 2/300
9/9 [==============================] - 0s 3ms/step - loss: 0.6079 - accuracy: 0.6620
Epoch 3/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5656 - accuracy: 0.5986
Epoch 4/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5605 - accuracy: 0.5845
Epoch 5/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5553 - accuracy: 0.5845
Epoch 6/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5506 - accuracy: 0.6127
Epoch 7/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5480 - accuracy: 0.6408
Epoch 8/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5424 - accuracy: 0.6338
Epoch 9/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5384 - accuracy: 0.6338
Epoch 10/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5372 - accuracy: 0.6479
Epoch 11/300
9/9 [===========

In [ ]:
ann8.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 0.7532 - accuracy: 0.8229


[0.7532340884208679, 0.8229166865348816]

####NN Nineth Attempt

I will try a different activation function in hidden layers

In [ ]:
ann9 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=64, activation='elu'),  # Increased neurons
    tf.keras.layers.Dense(units=24, activation='elu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer with 3 units for 3 classes
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann9.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann9.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=16, epochs=300, class_weight=class_weights)

Epoch 1/300
9/9 [==============================] - 1s 3ms/step - loss: 0.6887 - accuracy: 0.4789
Epoch 2/300
9/9 [==============================] - 0s 3ms/step - loss: 0.6055 - accuracy: 0.5634
Epoch 3/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5785 - accuracy: 0.5634
Epoch 4/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5657 - accuracy: 0.5563
Epoch 5/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5571 - accuracy: 0.6056
Epoch 6/300
9/9 [==============================] - 0s 2ms/step - loss: 0.5526 - accuracy: 0.6268
Epoch 7/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5453 - accuracy: 0.6408
Epoch 8/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5429 - accuracy: 0.6479
Epoch 9/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5393 - accuracy: 0.6479
Epoch 10/300
9/9 [==============================] - 0s 3ms/step - loss: 0.5359 - accuracy: 0.6408
Epoch 11/300
9/9 [===========

In [ ]:
ann9.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 5ms/step - loss: 0.9818 - accuracy: 0.7812


[0.9818222522735596, 0.78125]

Best Accuracy

####NN 10th Attempt

I will try dropouts to prevent overfitting in the more complex layer

In [ ]:
ann10 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=64, activation='relu'),  # Increased neurons
    tf.keras.layers.Dropout(0.3),  # Added dropout
    tf.keras.layers.Dense(units=24, activation='relu'),
    tf.keras.layers.Dense(units=1, activation='sigmoid')
])

In [ ]:
# Compile the model with the correct loss function for multi-class classification
ann10.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

ann10.fit(X_SelfVsInherited_train, y_SelfVsInherited_train, batch_size=32, epochs=300, class_weight=class_weights)

Epoch 1/300
5/5 [==============================] - 1s 4ms/step - loss: 0.7570 - accuracy: 0.5000
Epoch 2/300
5/5 [==============================] - 0s 6ms/step - loss: 0.7063 - accuracy: 0.5915
Epoch 3/300
5/5 [==============================] - 0s 5ms/step - loss: 0.7026 - accuracy: 0.6127
Epoch 4/300
5/5 [==============================] - 0s 3ms/step - loss: 0.6735 - accuracy: 0.6479
Epoch 5/300
5/5 [==============================] - 0s 3ms/step - loss: 0.6489 - accuracy: 0.7183
Epoch 6/300
5/5 [==============================] - 0s 5ms/step - loss: 0.6415 - accuracy: 0.7394
Epoch 7/300
5/5 [==============================] - 0s 4ms/step - loss: 0.6011 - accuracy: 0.7465
Epoch 8/300
5/5 [==============================] - 0s 3ms/step - loss: 0.5899 - accuracy: 0.7394
Epoch 9/300
5/5 [==============================] - 0s 4ms/step - loss: 0.6064 - accuracy: 0.7113
Epoch 10/300
5/5 [==============================] - 0s 4ms/step - loss: 0.6094 - accuracy: 0.7535
Epoch 11/300
5/5 [===========

In [ ]:
ann10.evaluate(X_SelfVsInherited_test,y_SelfVsInherited_test)

3/3 [==============================] - 0s 7ms/step - loss: 0.8310 - accuracy: 0.7812


[0.8309983611106873, 0.78125]

## 7. Evaluation and Comparison
Nearest Neighbor Classifier (NNC):
- Wealth Classification: Achieved an accuracy of 76.04%. This model had a fair performance but struggled with some imbalances in the dataset, as indicated by the confusion matrix.
- Self-Made vs. Inherited Wealth: Reached a higher accuracy of 83.33%. The model was more effective in this classification, possibly due to more distinct features distinguishing these categories.

Nearest Neighbor Classifier (NNC):
- Wealth Classification: Demonstrated a lower accuracy of 39.58%. This indicates that the centroid-based approach was less effective for this dataset, likely due to overlapping feature spaces among classes.
- Self-Made vs. Inherited Wealth: Improved accuracy to 67.71%, but still less effective compared to the NNC, suggesting that centroid-based classification is less suited for this data.

Neural Network (NN) Models:
- Wealth Classification: Varied accuracies (41.67% to 75.00%) across different attempts. The fluctuating performance might be due to overfitting or the complexity of the dataset. The highest accuracy (75.00%) indicates potential with further tuning.
- Self-Made vs. Inherited Wealth: More consistent performance, with accuracies ranging from 65.62% to 82.29%. This shows that NN models are relatively more effective for this classification task, with the highest accuracy models likely capturing the underlying patterns well.

Specifically, the Neural Network (NN) models were uniquely tailored to account for the imbalance within the dataset, a crucial step that was not mirrored in the other models used (such as the Nearest Neighbor Classifier and Nearest Class Centroid Classifier).

So NN models provide more trustworthy and balanced predictions
